### ISSUE TITLE
Ingest images and annotations from Data Drop 1

### ISSUE THREAD

**Issue by matt-bernstein** [ResonantGeoData/RGD-ScrumBoard#23]
# Context

IARPA have finalized [the format](https://infrastructure.smartgitlab.com/docs/pages/api_documentation.html#region-model) for releasing their annotations. The most recent batch are located here:

https://smartgitlab.com/TE/annotations/-/tree/master/D1/region_models

They are also mirrored to an s3 bucket with a few days' delay:

`s3://smart-imagery/annotations/D1/region_models`

The spacetime AOIs we need to ingest are laid out pretty transparently in these geojson files. You can get them as a list of geojson Features like so:

```python
import json
from glob import glob

regions = []
for fpath in sorted(glob('*.geojson')):
    with open(fpath) as f:
        annots = json.load(f)
        # the 'sites' are subsets of the 'regions', so we don't need them
        regions.extend([feat for feat in annots['features'] if feat['properties']['type'] == 'region'])
```

Example:
```
          {
            "type": "Feature",
            "properties": {
                "type": "region",
                "region_model_id": "US_Jacksonville_R01",
                "version": "1.0.0",
                "mgrs": "17RMP",
                "start_date": "2009-05-09",
                "end_date": "2020-01-26"
            },
            "geometry": {
                "type": "Polygon",
                "coordinates": [
                    [
                        [
                            -81.5968835679633,
                            30.2984987880008
                        ],
                        [
                            -81.5968835679633,
                            30.3667612161605
                        ],
                        [
                            -81.6953224473797,
                            30.3667612161605
                        ],
                        [
                            -81.6953224473797,
                            30.2984987880008
                        ],
                        [
                            -81.5968835679633,
                            30.2984987880008
                        ]
                    ]
                ]
            }
```
Example of 2 regions in their MGRS tile:
![52SDG](https://user-images.githubusercontent.com/60152561/122980354-8798e000-d366-11eb-8d83-3d56d9a5014f.png)

These regions will be continually updated outside of official data drop times (note the semver tag), to add new images and fix errors, though this should mostly affect the sites and not the regions. For example, D1 is apparently taking place on a rolling basis through August.

Future regions will be delivered in the folders `D2`, `D3`, etc, so we will be able to find them all in the same place.

# Data sources

Given the spacetime AOIs above, what data exactly do we need to ingest to cover each region?

## Annotations

In this project's nomenclature, a "region" is a subset of an MGRS spatial tile plus a time window. A "site" is a track of annotated polygons across several images plus its spatiotemporal bounding box; a subset of a "region". There can be multiple regions per tile and multiple sites per region. Think of a "region" as the thing to do Broad Area Search over to find "sites", which are construction sites. The time windows are as short as 1 year or as long as 2008-present.

First of all, it'd be nice to ingest the region and site annotations themselves for visualization purposes. (The site annotations are in the parallel `site_models` folder.) These have a formal spec linked above, and are essentially a set of nested, labeled, date-tagged GeoJSON Polygons and MultiPolygons in WGS84 coordinates.

## Annotated images

Each site annotation has a `source_name` tag which, glossing over the details, maps to a single LS, S2, or WV scene. The annotators are *not* using the same data sources as we are, and we have no access to their annotation/data processing pipeline, so this mapping step is inherently lossy.

The scenes can be found in the one of the following places:

1. `s3://smart-imagery/annotations/imagery/` is a bucket for IARPA to dump scenes that they think we will not be able to find anywhere else for one reason or another: they are at an earlier processing level (Real-Time as opposed to Tier 1), they are older versions that have since been overwritten in the public catalogs, or they are private/commercial/usage-restricted in some way. Scenes from all 3 sensors are here as one giant flat list of band files, many renamed, which includes panchromatic, RGB, and MSI scenes all together, and no ancillary files. This is like a temporary clearinghouse to get imagery out to us as fast as possible, before [for WV] ingesting it into their STAC catalog, or [for LS and S2] ???. *This is not recommended for ingestion into RGD,because it's far too unorganized and quickly changing to be relied on.*

2. For WV, the secondary source is hosted at `s3://smart-imagery/worldview-nitf/` and accessible as a STAC catalog through https://api.smart-stac.com/. *This was already ingested into the RGD WATCH instance, but it is being continually updated, so that ingestion should probably just be rerun on a schedule every week or so.*

3. For LS/S2, the secondary source is "wherever we can find it." Currently this is [Google Cloud](https://cloud.google.com/storage/docs/public-datasets), but I should do some experiments to see if [AWS](https://earth-search.aws.element84.com/v0) is a better source (see next point). *We can re-use our existing GCloud ingestion scripts for this, possibly with some looser restrictions (see next point).*

4. Nowhere/ambiguous. The following edge cases all exist in D1, and more will probably appear as more annotations are rolled out. 
    - A scene is not available anywhere, or cannot be identified from its `source_name`
    - [For WV] Only some spatial tiles of the scene are available.
    - [For WV] Only 8-channel MSI, 4-channel MSI, or PAN are available, not the full 9-channel MSI+PAN.
    - [For LS/S2] Only TCI is available, not MSI, and occasionally it's a tiny 700p thumbnail instead of the full resolution.
    - [For LS/S2] A scene is identified by a single datetime, and we don't know if it's the sensing datetime or the processing datetime, and both of those can be changed/duplicated as the GCloud and AWS catalogs are updated, so they may also be out of sync between the two catalogs.
    - [For LS] The annotated image is in Collection-2, and only Collection-1 is available in GCloud
    - [For S2] The annotated image is L2A, or L1C old-format, and we have to guess at the corresponding L1C new-format image on GCloud
 
In almost every case, we also don't know exactly which *band* or spatial resolution the annotators were looking at. This matters due to pixel-level shifts like rounding errors and parallax, since we have to project the annotations from geocoordinates back to pixel coordinates to work with them. In most cases it's possible to make an educated guess, though.

There are currently 513 annotated scenes covering 7 regions in D1. Roughly ~5% of annotated scenes are missing due to these issues and another ~10% may be unusable. This is small enough that we can just exclude these annotations for now, move on, and fix them later.

Note that ingestion into RGD is not a blocker for existing WATCH work, because the team already has access to the current version of D1 through DVC, which was done manually. Ingestion into RGD, as well as simplifying data access and allowing more powerful visualization tools, would act as a framework for future data drops to be handled more reliably than our current ~giant pile of python scripts~ process.

## Unannotated images

In addition to the annotated images, we'd also like to have access to all the other images spanning the region AOIs, for the same reason that we did in D0. This is a much easier task, because we don't need to get exactly the right version of each image, and deal with old-format/real-time/TCI-only and other special cases; we just want as many good MSI images as possible.

This is exactly the same as what we're already doing; we just need to reingest the WV STAC to get the updated version of it and reingest GCloud with the new D1 region AOIs instead of the D0 ones. So this is has by far the highest ratio of expected-value-to-effort.

# Tl;dr priority list

1. Ingest all LS/S2 from GCloud that cover the new D1 regions, and all of the updated WV STAC
2. Ingest D1 region and site annotation polygons
3. Where practical, ingest the specific version of each image that was actually annotated
4. Associate the annotations with their respective images, so each group is a single SpatialEntry, or something like that?
5. Continuous improvement of (3.) - add more data sources where necessary (AWS S2 L1C and LS L1C1, Sentinel-2 L2A, Landsat C2)
6. Automate the above process for future data drops, now that we know what they're going to look like

# Appendix

In future data drops, there may or may not be annotations on images in `s3://smart-imagery/planet/` as well.
